### PR TITLE
fix(files): Ignore included:false pattern

### LINKED ISSUE #1530: Karma 0.13 ignores included:false pattern
Hi. In one of my projects Karma started to include files marked with "included:false" in the karma.conf.js after I upgraded to Karma 0.13.3. Switching back to 0.12.37 resolves the issue so either there was a intended behavioral change (either undocumented or breaking some wrong assumptions on my side) or a regression. I reproduced the issue on Windows 10 as well as Ubuntu 15.04.

The karma.conf.js looks something like this:
```js
module.exports = function(config) {
  config.set({
    basePath: '../',
    ...
    files: [
      {pattern: 'src/scripts/backend/worker.js', included: false}, // Web-worker script with entry point
      'src/scripts/**/*.js',
      'test/unit/test_*.js'
    ],
    ...
  });
};
```
The worker.js script contains web-worker specific functionality and hence must not be executed directly.

Steps to reproduce:
* Download the reproducer from http://d0t.dbclan.de/dump/karma-included-false-ignored-in-0.13.zip and extract it somewhere
* Run ```npm install```
* Run ```npm test```
* Replace the package.json with package-working.json
* Re-run ```npm install```
* Re-run ```npm test```

**Expected:** As worker.js should not be included directly its contents should be ignored and the dummy test should pass in both runs of ```npm test```

**Actual:** For the first run with Karma 0.13.3 worker.js is included and Karma fails and complains about ```someWebworkerSpecificFunction``` in worker.js. The second run with Karma 0.12.37 works as expected.

As I said. I'm not sure if it's something I'm doing or not but the behavior definitely changed between those two versions.

### PR DESCRIPTION
**Pull request by budde377**
Do not include a file, when `include=false`.
Give `include=false` precedence over `include=true`.

Closes #1530

### CODE REVIEW

File: test/e2e/files.feature [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      PhantomJS
+      """
+
+  Scenario: Execute a test excluding an explicitly included file
+    Given a configuration with:
+      """
+      files = [
+        {pattern: 'files/log_foo.js', included: false},
+        {pattern: 'files/log_foo.js', included: true},
+        'files/*.js'
```
Comment (dignifiedquire) [inline test/e2e/files.feature]: It would be good to have one test case where `files/*.js` is the first entry.

**Comment by budde377** [comment]
@dignifiedquire There's an [integration test](https://github.com/karma-runner/integration-tests/blob/master/coverage-mocha-requirejs/karma.conf.js#L7) that fails, because it assumes some other semantic than changed precedence. Should the test be changed or should the semantics?

File: test/e2e/files.feature (hunk 1 above)
Comment (budde377) [inline test/e2e/files.feature]: :+1:

**Comment by dignifiedquire** [comment]
@budde377 right, so we have to think about this carefully, because in the 0.12 and before the semantics where that the most specific pattern would match, rather than `false > true`. So ideally I thought we could revert to that, what do you think?

**Comment by budde377** [comment]
@dignifiedquire. Lets do this the cool way! Inspired by the CSS specificity calculations, we should define a specificity-measure. I propose the following:

Minimatch supports:

 * Globstar
 * Wildcard
 * Extended glob matching
 * Brace expansion

Given a path, *p*, we can define a tuple *s(p) = (n<sub>g</sub>, n<sub>w</sub>, n<sub>e</sub>, n<sub>b</sub>)* where *n<sub>g</sub>*, *n<sub>w</sub>*, *n<sub>e</sub>*, and *n<sub>b</sub>* are the number of globstars, wildcards, extended glob match patterns, and brace expansion patterns, respectively. 

With any two paths, *p* and *q*, where *s(p)=(p<sub>1</sub>, p<sub>2</sub>, p<sub>3</sub>, p<sub>4</sub>)* and *s(q)=(q<sub>1</sub>, q<sub>2</sub>, q<sub>3</sub>, q<sub>4</sub>)* we can define a order of the corresponding tuples 

 *p<sub>1</sub> < p<sub>2</sub> ⇔ ∃ m > 0: ∀ i > m: p<sub>i</sub> = q<sub>i</sub> ∧ p<sub>m</sub> < q<sub>m</sub>* 

i.e. an lexicographical order.

Given two paths with different tuples, we can define the one with the smallest tuple (wrt. the order above) as the most specific. If the tuples are equal we can choose either one.

**Comment by budde377** [comment]
Or we could just let the number of files generated from a given pattern decide how specific it is. The more files, the less specific.

**Comment by dignifiedquire** [comment]
@budde377 :100: for the cool order you suggested above. We just need to make sure to document and test this well, so people understand what's happening :)

**Comment by budde377** [comment]
@dignifiedquire Yes. I'll start on it this afternoon.

**Comment by budde377** [comment]
After some failed attempts, I'm pretty sure that I have a solution. I'll have it implemented and tested primo next week.

**Comment by dignifiedquire** [comment]
Thanks @budde377 for the great work, looking forward to see it :)

**Comment by budde377** [comment]
@dignifiedquire I've now implemented an order. Please let me know if it needs more documentation etc.

**Comment by dignifiedquire** [comment]
Nice work, will review more closely tomorrow

File: test/unit/helper.spec.js [hunk 2]
```diff
@@ -245,4 +245,66 @@ describe('helper', () => {
...
+    it('should calculate right weight of pattern with glob star', () => {
+      helper.mmPatternWeight('**/a').should.be.deep.equal([1, 1, 0, 0, 0, 0])
+    })
+    it('should calculate right weight of pattern with misused glob star', () => {
+      helper.mmPatternWeight('***/a').should.be.deep.equal([1, 0, 3, 0, 0, 0])
+    })
+    it('should calculate right weight of pattern with more magic', () => {
+      helper.mmPatternWeight('{0..9}/?(a|b)c/[abc]/**/*.jsx?').should.be.deep.equal([10, 1, 1, 1, 1, 1])
+    })
+    it('should calculate right weight of pattern as worst glob set', () => {
```
Comment (dignifiedquire) [inline test/unit/helper.spec.js]: > worst glob set

:+1: :laughing:

**Comment by dignifiedquire** [comment]
Thanks @budde377 looking good, going to merge this and hope for the best :)

**Pull request merged**